### DISCUSSION TITLE
How do you set up a performant selector ( or in Vue's term, computed, something derived from multiple states) ?

### DISCUSSION THREAD

**Discussion by FateRiddle** [pmndrs/zustand#758] [Q&A] [answered]
What's the best way, performance wise, to do a selector?

By that I mean, if I've got `items` and `currentIndex` saved as zustand states, how does a component get access to "currentItem" which is defined by items[currentIndex] ?

If I fetch both `items` and `currentIndex` from useStore, then do the math in the given component, I'd imaging any change to `items` would result in a rerender of the component, yet the component is supposed to change only when "current item" is changed.

in code 
```js
export const useStore = create((set) => ({
  items: [5,6,3],
  currentIndex: 0,
}))
```

current item in this case is 5.

**Comment by dai-shi** [comment, marked as answer]
## Simple solution
This works pretty well. It's generally recommended until you hit any performance limit.
```js
const Component = () => {
  const currentItem = useStore((state) => state.items[state.currentIndex])
  // ...
}
```

## A little more optimization
With a stable selector, the library can do some more optimization. If the selector function is computation heavy, this may help.
```js
const selectCurrentItem = (state) => state.items[state.currentId]

const Component = () => {
  const currentItem = useStore(selectCurrentItem)
  // ...
}
```

## Memoizing selectors
If multiple components use the same selector **and** the selector function is computation heavy, you want to consider memoizing selectors. You can use `reselect` from `redux`, but my personal recommendation is [proxy-memoize](https://github.com/dai-shi/proxy-memoize).
```js
import memoize from 'proxy-memoize'

const selectCurrentItem = memoize((state) => state.items[state.currentId])

const Component = () => {
  const currentItem = useStore(selectCurrentItem)
  // ...
}
```

## computed properties
If you want computed properties in the store for some reason, like a vanilla-js usage, a middleware can be an option.
For example: https://github.com/cmlarsen/zustand-middleware-computed-state

**Reply by xiechao06** [reply]
Is subscribing to a state and computing the derived states an acceptable way? I think it may be a little more performant than using proxy-memoize, since it won't calculate the input object each time. Here is an example: https://codesandbox.io/s/zustand-derived-states-forked-0tluqw?file=/src/App.tsx

**Reply by dai-shi** [reply]
It's an okay way. There are some pitfalls, like useEffect is delayed a little bit and there's a chance count is updated before useEffect fires. If you want it a property on store, computing it on "set" would be better, like the middleware one.

I'm also experimenting a new way: https://github.com/dai-shi/derive-zustand

**Reply by xiechao06** [reply]
> computing it on "set" would be better

After some experiments， I found that using subscribing or computing it on set are all very hard to maintain, For example, if a derived state relies on states A, B, and C, one must compute it on "set" A, B, and C or subscribe to A, B, and C, now considering this state relies on one more state D, it is very easy to forget to compute it on "set" D or subscribe to D. 

So I think proxy-memoize is still a better way.
